File: solutions/lua/high-scores/1/high-scores.lua

---@class HighScores
local HighScores = {}
HighScores.__index = HighScores

function HighScores:scores()
	return self
end

function HighScores:latest()
	return self[#self]
end

function HighScores:personal_best()
	local best = 0
	for i = 1, #self do
		best = math.max(best, self[i])
	end
	return best
end

function HighScores:personal_top_three()
	local sorted = {}
	for i = 1, #self do
		table.insert(sorted, self[i])
	end
	table.sort(sorted, function(a, b) return a > b end)
	local top = {}
	for i = 1, 3 do
		table.insert(top, sorted[i])
	end
	return top
end

function HighScores.new(scores)
	local high_scores = setmetatable(scores, HighScores)
	return high_scores
end

return HighScores.new

File: solutions/zig/high-scores/1/high_scores.zig
const std = @import("std");

pub const HighScores = struct {
    scores: []const i32,
    // This struct, as well as its fields and methods, needs to be implemented.

    pub fn init(scores: []const i32) HighScores {
        return .{ .scores = scores };
    }

    pub fn latest(self: *const HighScores) ?i32 {
        return self.scores[self.scores.len-1];
    }

    pub fn personalBest(self: *const HighScores) ?i32 {
        var best: i32 = 0;
        for (self.scores) |score| {
            best = @max(best, score);
        }
        return best;
    }

    pub fn personalTopThree(self: *const HighScores) []const i32 {
        // std.mem.sort(i32, self.scores, {}, std.sort.asc(i32));
        return self.scores[self.scores.len-3..];
    }
};
